### PR TITLE
Thumbor 7

### PR DESCRIPTION
**Pull request by gingerlime**
* updated Dockerfile and docker-entrypoint based on
  https://github.com/MinimalCompact/thumbor/issues/67#issuecomment-592918001
* moving from Semaphore to Github actions
* using thumbor's built-in multiprocess
* taking changes from #101
* moving docker images to Github (ghcr.io)
* updated requirements with some tc (thumbor community) dependencies that support python 3 / thumbor 7
* added multiarch support (thanks @danquack and @mpdude)
* added option to specify THUMBOR_HOST (thanks @roman-stelmakh-emesa)

TODO

* [x] [circus / multiprocess is broken](https://github.com/MinimalCompact/thumbor/issues/67#issuecomment-592954554)
* [x] move dependencies to `requirements.txt` ?
* [x] add more libs to `requirements.txt` (which?)
* [x] is the conf file backwards compatible?
* [x] ModuleNotFoundError: No module named 'urllib2'
* [x] remotecv image isn't created/pushed? @danquack @mpdude ?
* [ ] caching/performance @danquack @mpdude 
* [ ] testing

### CODE REVIEW

**Comment by gingerlime** [comment]
@heynemann let's continue the work here? :)

I created a new thumbor:7.0.0 image on docker hub... I tried it and I'm getting this error however:

```
thumbor_1      | ---> Starting thumbor with 4 processes...
thumbor_1      | Traceback (most recent call last):
thumbor_1      |   File "/usr/local/bin/thumbor", line 8, in <module>
thumbor_1      |     sys.exit(main())
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/thumbor/server.py", line 142, in main
thumbor_1      |     importer = get_importer(config)
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/thumbor/server.py", line 61, in get_importer
thumbor_1      |     importer.import_modules()
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/thumbor/importer.py", line 83, in import_modules
thumbor_1      |     self.import_item("LOADER")
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/thumbor/importer.py", line 144, in import_item
thumbor_1      |     module = self.import_class(conf_value, get_module=True)
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/thumbor/importer.py", line 56, in import_class
thumbor_1      |     kls = import_class(name, get_module)
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/thumbor/importer.py", line 20, in import_class
thumbor_1      |     module = import_module(module_name)
thumbor_1      |   File "/usr/local/lib/python3.8/importlib/__init__.py", line 127, in import_module
thumbor_1      |     return _bootstrap._gcd_import(name[level:], package, level)
thumbor_1      |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
thumbor_1      |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
thumbor_1      |   File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
thumbor_1      |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
thumbor_1      |   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
thumbor_1      |   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
thumbor_1      |   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
thumbor_1      |   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
thumbor_1      |   File "<frozen importlib._bootstrap_external>", line 843, in exec_module
thumbor_1      |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
thumbor_1      |   File "/usr/local/lib/python3.8/site-packages/tc_aws/loaders/__init__.py", line 9, in <module>
thumbor_1      |     import urllib2
thumbor_1      | ModuleNotFoundError: No module named 'urllib2'
```

I guess we might need to add more libraries to requirements.txt?

File: thumbor/requirements.txt [hunk 1]
```diff
@@ -1,24 +1,11 @@
...
-botocore==1.12.253
-dateutils==0.6.6
-numpy==1.14.3
-remotecv==2.2.2
-opencv-engine==1.0.1
-thumbor==6.7.0
+remotecv==3.0.0
+boto==2.49.0
 tornado-botocore==1.5.0
-tc-aws==6.2.10
```
Comment (gingerlime) [inline thumbor/requirements.txt:12]: @heynemann I guess we still need some of the libraries here with thumbor 7 ? which ones? and which versions?

File: benchmarks/locust/Dockerfile [hunk 2]
```diff
@@ -1,6 +1,6 @@
-FROM python:3.6
+FROM python:3.8
```
Comment (gingerlime) [inline benchmarks/locust/Dockerfile:1]: which python version should we use?

File: thumbor/Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM python:2
+FROM python:3.8
```
Comment (gingerlime) [inline thumbor/Dockerfile:1]: which python version should we use?

Comment (heynemann) [reply thumbor/Dockerfile:1]: I'd go with 3.9, but any of 3.6+ will work (our CI tests those).

File: thumbor/requirements.txt (hunk 1 above)
Comment (heynemann) [reply thumbor/requirements.txt:12]: The problem with urllib2 is that tc-aws is not python3 compatible. https://stackoverflow.com/questions/2792650/import-error-no-module-name-urllib2
That said, I created a new aws storage library called thumbor-aws until tc-aws gets updated. We can try that... https://github.com/thumbor/thumbor-aws

Comment (heynemann) [reply thumbor/requirements.txt:12]: I created a new library called thumbor-aws we can use until tc-aws gets updated to python3. https://github.com/thumbor/thumbor-aws

The problem with urllib2 is that tc-aws is not python3 compliant. https://stackoverflow.com/questions/2792650/import-error-no-module-name-urllib2

**Comment by heynemann** [comment]
Regarding the conf file it's super fine. We didn't change any configuration during our "revamp". Added some as the handlers list, but configurations remain the same.

File: thumbor/requirements.txt (hunk 1 above)
Comment (gingerlime) [reply thumbor/requirements.txt:12]: so should we add thumbor-aws to requirements.txt ?

Comment (gingerlime) [reply thumbor/requirements.txt:12]: also, correct me if I'm wrong, but wouldn't this require a change to the configs? e.g. if someone uses `LOADER=tc_aws.loaders.s3_loaders` they would need to change it? if so, is there a way to make thumbor-aws backwards compatible, so it accepts the "old" config directives?

Comment (heynemann) [reply thumbor/requirements.txt:12]: Yep, I meant as an additional lib to the docker image :) That way people who want can use it while tc-aws is not compatible.

Comment (heynemann) [reply thumbor/requirements.txt:12]: There is :) will work on making it more compatible later...

Comment (gingerlime) [reply thumbor/requirements.txt:12]: Great. What's the next step here then? I'm not sure I completely follow :)

File: thumbor/requirements.txt [hunk 4]
```diff
@@ -1,26 +1,11 @@
...
-tc-redis==1.0.1
-thumbor-cloud-storage==3.0.0
-pgmagick==0.7.4
-graphicsmagick-engine==0.1.1
+python-dateutil
+dateutils==0.6.12
+shortuuid==1.0.8
+redis==3.5.3
+raven==6.10.0
 cairosvg==1.0.22
```
Comment (gingerlime) [inline thumbor/requirements.txt:10]: can we update cairosvg to `CairoSVG>=2.5.1` ? it seems like this version has a security vulnerability?

Comment (heynemann) [reply thumbor/requirements.txt:10]: I guess... We can definitely try it.

**Comment by heynemann** [comment]
Sorry for not being super active here. Was on a vacation with my family. 🥥 🌴 🍹

**Comment by mpdude** [comment]
What kind of help would be required here to bring this forward?

**Comment by gingerlime** [comment]
Good question @mpdude. I'm not sure I have the full picture either.

There are a couple of checklist items on this PR, and probably more testing involved... I mostly rely on @heynemann and @kkopachev from the thumbor project to be honest, because I'm not that familiar with the internals / dependencies / deployment requirements of the project, which we need to iron-out in order to get a new docker image available... Hopefully we won't break backwards compatibility for people who are using thumbor 6, or try to minimize breakage and provide clear upgrade path, documentation, limitations etc.

**Comment by maximka777** [comment]
Faced an issue running docker container based on image built in the branch:
```
Traceback (most recent call last):
  File "/usr/local/bin/envtpl", line 5, in <module>
    from envtpl import main
  File "/usr/local/lib/python3.9/site-packages/envtpl.py", line 150, in <module>
    @jinja2.contextfunction
AttributeError: module 'jinja2' has no attribute 'contextfunction'
---> Starting thumbor with 1 processes...
Traceback (most recent call last):
  File "/usr/local/bin/thumbor", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/thumbor/server.py", line 135, in main
    config = get_config(
  File "/usr/local/lib/python3.9/site-packages/thumbor/server.py", line 45, in get_config
    return Config.load(config_path, conf_name="thumbor.conf", lookup_paths=lookup_paths)
  File "/usr/local/lib/python3.9/site-packages/derpconf/config.py", line 75, in load
    raise ConfigurationError('Configuration file not found at path %s' % path)
derpconf.config.ConfigurationError: Configuration file not found at path /app/thumbor.conf
```

Used commands:
- to build `docker build -t minimal-compact-thumbor-7:local.1 .` (in `./thumbor`)
- to run `docker run -p 80:80 minimal-compact-thumbor-7:local.1`

After that, I tried to manually create and copy a config file into image (in `Dockerfile`), and it worked nicely.
I will try to find a solution for the issue, but leave it here in case anybody of you have ideas.

**Comment by mpdude** [comment]
Does anyone have a clear picture/idea of what we need to do to get this resolved?

**Comment by gingerlime** [comment]
@mpdude see https://github.com/MinimalCompact/thumbor/pull/103#issuecomment-1084129918 ? there's a task list on the PR description. There's also #116 which could be related to your question?

**Comment by mpdude** [comment]
I've been able to (cross-)build at least the basic thumbor image with the configuration from this branch using Docker Desktop for Mac, and it works on x86_64 hardware. A co-worker of mine successfully built and ran it on an M1 chip.

So – what is this urllib checklist item about, and how can we find out if/whether any additional requirements are missing?

File: thumbor/Dockerfile [hunk 5]
```diff
@@ -1,4 +1,4 @@
-FROM python:2.7.18-buster
+FROM python:3.9
```
Comment (mpdude) [inline thumbor/Dockerfile:1]: Maybe `python:3.11-bullseye` would be a more up-to-date alternative?

File: thumbor/requirements.txt (hunk 1 above)
Comment (mpdude) [reply thumbor/requirements.txt]: @heynemann is all this resolved since tc_aws 7.0.2 works with Thumbor 7?

Comment (gingerlime) [reply thumbor/requirements.txt]: I updated tc-aws to 7.0.2. There's a new thumbor 7.4.7 image on docker hub. Can you test it? ideally with tc-aws. I think this should solve the urllib2 issue.

**Comment by gingerlime** [comment]
@mpdude take a look at #121. Can you maybe help there?

**Comment by mpdude** [comment]
Am I right assuming that Circus for running Thumbor multi-process has been dropped, since that is built into Thumbor 7? 

If yes: It seems there are some configuration left-overs that we could remove.

**Comment by gingerlime** [comment]
@mpdude yes, circus is no longer needed. Thumbor can run multi-process. If you can you open a PR to get rid of any left-overs that you find it would be great 👍

**Comment by gingerlime** [comment]
@danquack @mpdude I think we're getting closer to getting this out. A couple of things I could really use your help on though:

1. Testing. I don't have access to a realistic environment, so would be great if you can test it somewhere (ideally some kind of staging environment)
2. can we improve caching/performance? the build / push is super slow, I guess multiarch takes longer? I tried using `gha` caching, but it doesn't feel like it has an effect. Maybe because most builds failed so far? I'm not sure.

**Comment by gingerlime** [comment]
Somehow the remotecv image isn't created/pushed. I cannot see any errors however, so it's weird. @danquack can you take a look?

**Comment by gingerlime** [comment]
I'll go ahead and merge these changes and move to Thumbor 7. We can deal with any issues later, and there's still the old `latest` and `6.x` tags on docker hub anyway.

**Pull request merged**